### ISSUE TITLE
Cork Library for Boolean CSG

### ISSUE THREAD

**Issue by drlukeparry** [mikedh/trimesh#938]
For my own project I have considered using Cork CSG library, which compiles with minimal dependencies. Despite its age it seems to be reliable and respectable performance. Was there a particular reason for not choosing this as an engine backend?

I am also looking at the following fast-csg implementation

https://github.com/sallesviana/MeshIntersection

**Comment by mikedh** [comment]
Agreed, `trimesh` could definitely use a better Boolean engine option and I'm open to suggestions!

I looked at Cork a few years ago, the reason I didn't add it as an engine option was that it wasn't actively maintained and in my brief testing (this may not be fair in 2020) my impression was that Cork was less robust than than SCAD and slower than Blender. 

"Maintained" can be kind of arbitrary but it would be nice to see 1 or more commits in the last year and packaged for major platforms (ideally using something like [cibuildwheel](https://github.com/joerick/cibuildwheel)). Seems like the "dream" dependency here would be a tidy c++ implementation with nice Python bindings that outperforms both Blender and SCAD. If that existed we can build wheels for major platforms relatively easily (aka [triangle](https://github.com/drufat/triangle)).

Last time I looked into this I was impressed by the simplicity of [csg.js](https://evanw.github.io/csg.js/) which generated a number of [c++](https://github.com/dabroz/csgjs-cpp) and [python](https://github.com/timknip/pycsg) implementations. Fast-CSG looks interesting as well! 

I think the first step is probably a test-validate-profile suite which includes the major edge cases. Here's a first pass (using [this model](https://github.com/mikedh/trimesh/blob/fix/svgmeta/models/boolean.glb)):

```
import trimesh
from pyinstrument import Profiler

if __name__ == '__main__':
    # a simple scene with one base and a bunch of cuts
    s = trimesh.load('models/boolean.glb')
    base = s.geometry['base']
    cuts = [v for k, v in s.geometry.items() if k.startswith('cut')]

    print('\n\nSCAD\n')
    profiler = Profiler()
    profiler.start()

    a = base.difference(cuts, engine='scad')
    b = base.union(cuts, engine='scad')
    # note : crashes SCAD -_-
    #c = base.intersection(cuts, engine='scad')

    profiler.stop()
    print(profiler.output_text(unicode=True, color=True))

    print('\n\nBLENDER\n')
    profiler = Profiler()
    profiler.start()

    a = base.difference(cuts, engine='blender')
    b = base.union(cuts, engine='blender')

    profiler.stop()
    print(profiler.output_text(unicode=True, color=True))
``` 

Returns the result that aligns with my existing impression (SCAD is slow, blender is fast but explodes randomly): 


```
SCAD


  _     ._   __/__   _ _  _  _ _/_   Recorded: 13:00:39  Samples:  64
 /_//_/// /_\ / //_// / //_'/ //     Duration: 20.279    CPU time: 0.090
/   _/                      v3.0.1

Program: /dropbox/drop/Dropbox/robotics/trimesh/prof.py

20.278 <module>  prof.py:1
├─ 14.054 difference  trimesh/base.py:2603
│  └─ 14.054 difference  trimesh/boolean.py:10
│     └─ 14.054 boolean  trimesh/interfaces/scad.py:57
│        └─ 14.054 interface_scad  trimesh/interfaces/scad.py:34
│           └─ 14.017 run  trimesh/interfaces/generic.py:69
│              └─ 13.999 check_call  subprocess.py:276
│                    [8 frames hidden]  subprocess
│                       13.995 _try_wait  subprocess.py:1401
└─ 6.224 union  trimesh/base.py:2581
   └─ 6.224 union  trimesh/boolean.py:29
      └─ 6.224 boolean  trimesh/interfaces/scad.py:57
         └─ 6.224 interface_scad  trimesh/interfaces/scad.py:34
            └─ 6.203 run  trimesh/interfaces/generic.py:69
               └─ 6.197 check_call  subprocess.py:276
                     [8 frames hidden]  subprocess
                        6.191 _try_wait  subprocess.py:1401




BLENDER

face_normals incorrect shape, ignoring!

  _     ._   __/__   _ _  _  _ _/_   Recorded: 13:00:59  Samples:  26
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.749     CPU time: 0.038
/   _/                      v3.0.1

Program: /dropbox/drop/Dropbox/robotics/trimesh/prof.py

1.748 <module>  prof.py:1
├─ 0.959 union  trimesh/base.py:2581
│  └─ 0.959 union  trimesh/boolean.py:29
│     └─ 0.959 boolean  trimesh/interfaces/blender.py:40
│        └─ 0.948 run  trimesh/interfaces/generic.py:69
│           └─ 0.945 check_call  subprocess.py:276
│                 [8 frames hidden]  subprocess
│                    0.941 _try_wait  subprocess.py:1401
└─ 0.788 difference  trimesh/base.py:2603
   └─ 0.788 difference  trimesh/boolean.py:10
      └─ 0.788 boolean  trimesh/interfaces/blender.py:40
         └─ 0.771 run  trimesh/interfaces/generic.py:69
            └─ 0.771 check_call  subprocess.py:276
                  [8 frames hidden]  subprocess
                     0.767 _try_wait  subprocess.py:1401

```

**Comment by drlukeparry** [comment]
Yes I can understand why you have gone with this route. OpenSCAD is using CGAL Nef implementation which from what I briefly read is accurate but very slow.

With Blender, I am not sure whether they transitioned to their own implementation with 2.8 release. I have tried with 2,6 for 3D printing and found it to unrelaible for use for creating manifold meshes in 3D printing. I generally don't like the idea of  linking to blender because of its portability. I quickly found the following development log looking into this further as a side note:

https://developer.blender.org/T67744 

Having a look there is also fastcsg which was developed by IRIA but is not opensource (guessing that doesn't matter).

http://morpheo.inrialpes.fr/static/QuickCSG/

Some method don't work for non-robust non-manifold models - this could be internally checked as an extra option perhaps? 
I suppose given that you have the flexibility for different engines, we are not fixed in other alternatives - and people can choose based on how robust/quick they desire their operations.

**Comment by mikedh** [comment]
Quick update:
- I got `meshIntersection` to compile, it needs some cleanup but looks pretty good generally. I *did* notice however that it only handles intersection, not differences or unions. 
- I started on some python bindings for `csgjs_cpp` . This may be too simplistic to handle the precision issues properly (what people seem to struggle with).
- `FastCSG` looks pretty solid but it not being open-source is unfortunate and probably a deal breaker for being a dependency. 
- I should probably benchmark `cork` again

**Comment by AgileEnabler** [comment]
Have you looked at using CGAL directly for the boolean operations? I found it to be fast and robust (for watertight meshes).  However, I have not yet found a comfortable Python binding that doesn't somehow required root access.  Maybe https://github.com/nschloe has a bit of experience in how to do that...?

https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__corefinement__grp.html

**Comment by mikedh** [comment]
Hey, I guess I assumed CGAL in terms of speed would be (OpenSCAD - Import/Export Overhead), where import and export overhead is quite low.

**Comment by rujialiu** [comment]
In case you're still interested, QuickCSG is now opensource (https://gitlab.inria.fr/quickcsg/quickcsg) @mikedh
